src/compiler/checker.ts

```diff
@@ -25011,7 +25011,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const constraint = getConstraintOfTypeParameter(inference.typeParameter);
             if (constraint) {
                 const instantiatedConstraint = instantiateType(constraint, context.nonFixingMapper);
-                if (!inferredType || !context.compareTypes(inferredType, getTypeWithThisArgument(instantiatedConstraint, inferredType))) {
+                if (!inferredType || inferredType === silentNeverType || !context.compareTypes(inferredType, getTypeWithThisArgument(instantiatedConstraint, inferredType))) {

```

Comment:
I tried to manufacture this test case but I don't have a good grasp of the problem and I failed to do so.

Reply:
Something like `type Cond<T> = T extends (a: infer A extends B, b: infer B) => any ? [A, B] : never` would have it's behavior changed by this, I think, I'm not wholly sure, either - it's just the only kind of construct I can think of where inference uses the `NoDefaults` flag and one type parameter's inference may depend on another one.

Reply:
> it's just the only kind of construct I can think of where inference uses the NoDefaults flag and one type parameter's inference may depend on another one.

I think that this construct might be using a different way of suppressing the defaults or something. If I understand correctly then `InferenceFlags.NoDefaults` is only used when cloning the outer context in situations when the compiler has to deal with nested inference contexts.

Reply:
So it is! Dunno why I thought it was used for `extends` clause inference. I guess that means you'd need to be resolving a nested call where the inner call relies on the type parameters from the outer call, perhaps obviously, similar to the new test case.

I think I know why this seems so funky to me now, having looked at it. This has the effect of just removing the `silentNeverType` from being set as an inference result entirely. It's only set a few lines above this in the `context.flags & InferenceFlags.NoDefault` case, and then down here, now, after explicitly setting it, so long as the type parameter has a constraint, we're basically undoing it. This change would be the same as making that branch
```ts
                else if (context.flags & InferenceFlags.NoDefault) {
                    // We use silentNeverType as the wildcard that signals no inferences.
                    inferredType = getConstraintOfTypeParameter(inference.typeParameter) ? instantiateType(getConstraintOfTypeParameter(inference.typeParameter), context.nonFixingMapper) : silentNeverType;
                }
```
except it also nukes silent nevers which actually came from not-just-this-type-parameter's-lack-of-an-inference source, too.

Now, AFAIK, the point of the `NoDefault` `InferenceFlag` was expressly to make it so we didn't infer a default or constraint for a type parameter when we still had a non-return-type pass to get a better result from, since the return type mapper gets is used to inform the contextual types for the inference context to produce a result. I think what we'd actually want to do is somehow see if we actually produce no further results for the given type parameter and then, if so, yield the constraint, since basically we just never want to surface the `never` as the final inference result for the actual top-level inference context.

Taking all that together, I _think_ what we actually wanna do to accomplish what we want here is to change `cloneInferredPartOfContext` so it no longer considers a type parameter with a `silentNeverType` `inferredType` as an inferred part of the context, since, definitionally, it's our result for no-inference-was-made! Doing so should prevent the not-a-result result from flowing into the actual outer inference context.

Reply:
The problem here is not about the outer context though. It's more about the inner one - the `inferredType` there isn't "informed" by the contextual type of the outer one. Perhaps the inferred type from the inner one can inform the outer one and that can inform the inner one in an additional pass or something, I don't know. Please also bear in mind that I only understand scraps of the overall algorithm, the bigger picture often escapes me as I just learn how this all works as I go and debug stuff.

I wonder if you have seen my analysis of what currently happens which can be found at the bottom of my post [here](https://github.com/microsoft/TypeScript/issues/52864)? I think that perhaps that could help you understand which part of the algorithm is at fault here.

Note that from my limited knowledge about this, it's the `instantiatedType` of the `contextualType` that is not correct. There are no inferences made from this `silentNeverType` to whatever we currently infer to (for example: `Matcher<input>`). `inferFromTypes` just ignores a `source` that is `never`.

It's also worth noting that at some point in this the `silentNeverType` is **observable** in the userland, we can "capture"/observe it with types like `IsNever`, see the example [here](https://github.com/gvergnaud/ts-pattern/commit/12aa9d38e935e181d8e49418d0df59742bf9a3ff#diff-2a614f2c1e3a277d612a1442290ba3d24933207799bbcd8ed2fa887e8a8e299fR40-R48). I think that this is a problem as well as that implementation detail shouldn't be surfaced this way.

Given that this `silentNeverType` doesn't really spread into outer inferences **in this situation**, the proposed change to `cloneInferredPartOfContext` doesn't affect this whole thing anyhow. I tried this patch:
```diff
diff --git a/src/compiler/checker.ts b/src/compiler/checker.ts
index 2f9ad3e330..98d1e7e2b3 100644
--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -23884,7 +23884,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function cloneInferredPartOfContext(context: InferenceContext): InferenceContext | undefined {
-        const inferences = filter(context.inferences, hasInferenceCandidates);
+        const inferences = filter(context.inferences, inference => hasInferenceCandidates(inference) && inference.inferredType !== silentNeverType);
         return inferences.length ?
             createInferenceContextWorker(map(inferences, cloneInferenceInfo), context.signature, context.flags, context.compareTypes) :
             undefined;
```
It had no negative effect on the existing test suite. Maybe there is some other test to be found that would prove the need for this change.